cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -419,11 +430,15 @@ public ApplicationMeta getLatest(ApplicationReference appReference) throws IOExc
   }
 
   public List<ApplicationMeta> getAllAppVersions(ApplicationReference appRef) throws IOException {
```

Comment:
unrelated, but is there a jira open to improve this so that we don't load all versions into memory? Ideally we would return the iterator, so that the results can be streamed through, but at the very least there should be a limit. If not, we should open a jira, as this will eventually result in out of memory errors once somebody has enough versions.

Reply:
This is a good point, ticket created: https://cdap.atlassian.net/browse/CDAP-20166

Reply:
Note that returning an iterator is wrong as it's easy to break transaction boundaries. See io.cdap.cdap.app.store.Store#scanApplications for a functional way to do it